api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

```diff
@@ -394,34 +394,22 @@ public boolean isSecretAnswer(User u, String answer) {
 	 * @param new user that has privileges
 	 */
 	private void checkPrivileges(User user) {
-		Collection<Role> roles = user.getAllRoles();
-		List<String> requiredPrivs = new ArrayList<>();
-		
-		for (Role r : roles) {
-			if (r.getRole().equals(RoleConstants.SUPERUSER)
-			        && !Context.hasPrivilege(PrivilegeConstants.ASSIGN_SYSTEM_DEVELOPER_ROLE)) {
-				throw new APIException("User.you.must.have.role", new Object[] { RoleConstants.SUPERUSER });
-			}
-			if (r.getPrivileges() != null) {
-				for (Privilege p : r.getPrivileges()) {
-					if (!Context.hasPrivilege(p.getPrivilege())) {
-						requiredPrivs.add(p.getPrivilege());
-					}
-				}
-			}
-		}
-		
+		List<String> requiredPrivs = user.getAllRoles().stream().peek(this::checkSuperUserPrivilege)
+				.map(Role::getPrivileges).filter(Objects::nonNull).flatMap(Collection::stream)
+				.map(Privilege::getPrivilege).filter(p -> !Context.hasPrivilege(p)).collect(Collectors.toList());
 		if (requiredPrivs.size() == 1) {
 			throw new APIException("User.you.must.have.privilege", new Object[] { requiredPrivs.get(0) });
 		} else if (requiredPrivs.size() > 1) {
-			StringBuilder txt = new StringBuilder("You must have the following privileges in order to assign them: ");
-			for (String s : requiredPrivs) {
-				txt.append(s).append(", ");
-			}
-			throw new APIException(txt.substring(0, txt.length() - 2));
+			throw new APIException(String.format(ERROR_FORMAT, String.join(", ", requiredPrivs)));
```

Comment:
@sherrif10 you might have to i18n this `ERROR_FORMAT` message. Like it is done here: https://github.com/openmrs/openmrs-core/pull/3163/files#diff-38273d709321a17f2a9eabb804784c82R410

Reply:
Thanks but , i extracted this ERROR_FORMAT message and gave its own string, is there any modification that is needed

Reply:
Unfortunately, i have not created unit tests since there are similar same testing that tests the same functionality in  **userServiceTest class.**

Reply:
i think we can merge this ticket and get back other reference tickets ,unless other requirements needed .cc @dkayiwa  @ibacher

Reply:
@sherrif10 what about my two comments above?
- i18n of your `ERROR_FORMAT` message (the naming of this variable is not ok btw.)
- a test that shows that `peek(..)` work as expected to throw that exception.

Reply:
ohh sorry let me try creating a test for that. Any suggestion of changing the  ERROR_FORMAT message

Reply:
@sherrif10 did you see this? https://github.com/openmrs/openmrs-core/pull/3163#discussion_r402342472

I pointed to an example somewhere else in the same file that you modified.

Reply:
ohh yes i have seen thanks for that pointer let me do it